src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -324,70 +327,41 @@ double SolverInterfaceImpl::initialize()
   double time       = 0.0;
   int    timeWindow = 1;
 
-  PRECICE_DEBUG("Initialize coupling schemes");
-  _couplingScheme->initialize(time, timeWindow);
-  PRECICE_ASSERT(_couplingScheme->isInitialized());
-
-  double dt = _couplingScheme->getNextTimestepMaxLength();
-
   for (auto &context : _accessor->readDataContexts()) {
     context.initializeWaveform();
   }
 
-  if (_couplingScheme->hasDataBeenReceived()) {
-    performDataActions({action::Action::READ_MAPPING_PRIOR}, 0.0, 0.0, 0.0, dt);
-    mapReadData();
-    performDataActions({action::Action::READ_MAPPING_POST}, 0.0, 0.0, 0.0, dt);
-  }
-
-  PRECICE_INFO(_couplingScheme->printCouplingState());
-
-  solverInitEvent.start(precice::syncMode);
-
   _meshLock.lockAll();
 
-  _state = State::Initialized;
-
-  return _couplingScheme->getNextTimestepMaxLength();
-}
-
-void SolverInterfaceImpl::initializeData()
-{
-  PRECICE_TRACE();
-  PRECICE_CHECK(!_hasInitializedData, "initializeData() may only be called once.");
-  PRECICE_CHECK(_state != State::Finalized, "initializeData() cannot be called after finalize().")
-  PRECICE_CHECK(_state == State::Initialized, "initialize() has to be called before initializeData()");
-  PRECICE_ASSERT(_couplingScheme->isInitialized());
-  PRECICE_CHECK(not(_couplingScheme->sendsInitializedData() && isActionRequired(constants::actionWriteInitialData())),
-                "Initial data has to be written to preCICE by calling an appropriate write...Data() function before calling initializeData(). "
-                "Did you forget to call markActionFulfilled(precice::constants::actionWriteInitialData()) after writing initial data?");
-
-  auto &solverInitEvent = EventRegistry::instance().getStoredEvent("solver.initialize");
-  solverInitEvent.pause(precice::syncMode);
-
-  Event                    e("initializeData", precice::syncMode);
-  utils::ScopedEventPrefix sep("initializeData/");
-
-  PRECICE_DEBUG("Initialize data");
   double dt = _couplingScheme->getNextTimestepMaxLength();
 
   performDataActions({action::Action::WRITE_MAPPING_PRIOR}, 0.0, 0.0, 0.0, dt);
   mapWrittenData();
   performDataActions({action::Action::WRITE_MAPPING_POST}, 0.0, 0.0, 0.0, dt);
 
-  _couplingScheme->initializeData();
+  PRECICE_DEBUG("Initialize coupling schemes");
+  _couplingScheme->initialize(time, timeWindow);
+  dt = _couplingScheme->getNextTimestepMaxLength(); // dt can change at this point again, if second participant and participant-first method for setting time window size is used.
```

Comment:
Does `dt` for the write mapping of initialize data even make sense at all?

At this point, there hasn't been any time-step and the time-window size isn't known either. 

Why don't we simply pass `0.0` or an invalid value like `-1` instead?
This would allow the user to distinguish between normal and init data actions too.
This would simplify some things.

Reply:
I'm not very deep in the "actions-business". Currently there is clearly a design issue, if write actions are used with the participant first method, because the second participant just cannot know the `dt` at this point in time. If `dt` is given in the config - and therefore fixed - there is no problem. I'll change it to `0.0` for now, feel free to change or comment, if something else looks more reasonable.

Reply:
I changed the `dt` to `0.0`. This is dangerous, because of the `ScaleByDt` action, which divides by `dt`...
I also opened an issue #1358.

Reply:
I think we can simply remove the argument, see my comment in #1358.
So "0.0" is OK here. Both "breaks" should go into v3.0 anyway.

Reply:
Added some comments here: 6865868d0061d632e3cc1846bbf1a46e60fa814f.